### PR TITLE
Make the installer from the JavaWrapper instead of using launch4j

### PR DESCRIPTION
**Pull request by 0xilly**
This PR depends on the java wrapper being deployed to forge maven

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -197,25 +198,138 @@ tasks.build.dependsOn makeMdk
...
+    extension = 'tar.gz'
+    destinationDir file("$buildDir/distributions/")
+    compression = Compression.GZIP
+    from "$buildDir/jwrappers/"
+    include "forge-$project.version-installer-linux"
+}
+
+task pkgOsx(type: Tar) {
+    dependsOn renameWrappers
+    classifier = "osx"
```
Comment (kashike) [inline build.gradle]: Should probably be "mac" or "macos", same for the task name and other occurrences of "osx".

**Comment by mezz** [comment]
Can you explain the purpose of the PR for people out of the loop?

File: jsons/1.11.2.json [hunk 2]
```diff
@@ -558,6 +558,6 @@
     "minecraftArguments": "--username ${auth_player_name} --version ${version_name} --gameDir ${game_directory} --assetsDir ${assets_root} --assetIndex ${assets_index_name} --uuid ${auth_uuid} --accessToken ${auth_access_token} --userType ${user_type} --versionType ${version_type}",
     "minimumLauncherVersion": 18,
     "releaseTime": "2016-12-21T09:29:12+00:00",
-    "time": "2017-01-12T10:44:50+00:00",
+    "time": "2017-02-07T13:18:39+00:00",
```
Comment (liach) [inline jsons/1.11.2.json]: Why to change this file?

Comment (mezz) [reply jsons/1.11.2.json]: Every time you set up the project on 1.11.2 it updates this file, they probably commited it by accident.  
It should probably be git-ignored or updated, either way not a big deal as far as I know.

Comment (0xilly) [reply jsons/1.11.2.json]: I honestly have no idea how those changed

Comment (LexManos) [reply jsons/1.11.2.json]: Its specifically NOT git ignored so we know when it changes. So we can react to any changes Mojang does to the launcher.

File: build.gradle (hunk 1 above)
Comment (0xilly) [reply build.gradle]: hmm I realize im going to have to package a app dir for osx so I'll change then

Side Note this is still a better then the original name darwin

Comment (LexManos) [reply build.gradle]: actually its a fairly common practice to call it OSX as its not compatible with other mac OS's like PPC. So OSX is fine.

File: jsons/1.11.2.json (hunk 2 above)
Comment (mezz) [reply jsons/1.11.2.json]: Should this file be removed from the PR?

Comment (LexManos) [reply jsons/1.11.2.json]: meh doesn't matter. As long as I notice it when pulling in the PR.
There isn't any functional change we need to care about.

File: build.gradle (hunk 1 above)
Comment (kashike) [reply build.gradle]: I'd agree, except Apple re-branded OSX as macOS.
https://en.wikipedia.org/wiki/MacOS

Comment (LexManos) [reply build.gradle]: Ya, don't care for them re-branding. It's still OSX.

File: build.gradle [hunk 3]
```diff
@@ -197,25 +198,138 @@ tasks.build.dependsOn makeMdk
...
+def wrapOsxPath = file("$buildDir/jwrappers/$wrapOsx")
+def wrapWinPath = file("$buildDir/jwrappers/$wrapWin")
+
+task wrapInstaller(type: Copy) {
+    dependsOn tasks.installer, configurations.wrapTar, configurations.wrapZip
+
+    from zipTree(configurations.wrapZip.singleFile)
+    into "$buildDir/jwrappers"
+
+    from {
```
Comment (mattmess1221) [inline build.gradle:233]: probably don't need this `from` block here.

File: build.gradle [hunk 4]
```diff
@@ -197,25 +198,138 @@ tasks.build.dependsOn makeMdk
...
+        configurations.wrapTar.asFileTree.each {
+            from tarTree(it)
+        }
+        null
```
Comment (mattmess1221) [inline build.gradle]: unnecessary return

File: build.gradle [hunk 5]
```diff
@@ -197,25 +198,138 @@ tasks.build.dependsOn makeMdk
...
+    doLast {
+
+        copy {
+            from("$buildDir/jwrappers/")
+            into("$buildDir/jwrappers")
+            rename("javawrapper-v$wrapper_version-linux", "forge-$project.version-installer-linux")
+            rename("javawrapper-v$wrapper_version-osx", "forge-$project.version-installer-osx")
+            rename("javawrapper-v$wrapper_version-win.exe", "forge-$project.version-installer-win.exe")
+        }
+        delete {
```
Comment (mattmess1221) [inline build.gradle]: Unnecessary `delete` block

File: build.gradle [hunk 6]
```diff
@@ -197,25 +198,138 @@ tasks.build.dependsOn makeMdk
...
+
+tasks.build.dependsOn 'pkgWrapped'
 
 // MAVEN
 
 artifacts {
     if (project.hasProperty('forgeJenkinsPass'))
         archives changelog.output
-    archives file("build/distributions/${project.name}-${project.version}-installer-win.exe")
+//    archives file("build/distributions/${project.name}-${project.version}-installer-win.exe")
```
Comment (mattmess1221) [inline build.gradle]: Should probably delete this instead of commenting it out.

File: build.gradle [hunk 7]
```diff
@@ -197,25 +198,138 @@ tasks.build.dependsOn makeMdk
...
+task renameWrappers() {
+    dependsOn wrapInstaller
+    doLast {
+
+        copy {
+            from("$buildDir/jwrappers/")
+            into("$buildDir/jwrappers")
+            rename("javawrapper-v$wrapper_version-linux", "forge-$project.version-installer-linux")
+            rename("javawrapper-v$wrapper_version-osx", "forge-$project.version-installer-osx")
+            rename("javawrapper-v$wrapper_version-win.exe", "forge-$project.version-installer-win.exe")
```
Comment (mattmess1221) [inline build.gradle:261]: `rename "^javawrapper-v$wrapper_version-(linux|osx|win\\.exe)\$", "forge-$project.version-installer-\$1"`
That should cover all 3.

File: build.gradle [hunk 8]
```diff
@@ -197,25 +198,138 @@ tasks.build.dependsOn makeMdk
...
+
+def wrapLinuxPath = file("$buildDir/jwrappers/$wrapLinux")
+def wrapOsxPath = file("$buildDir/jwrappers/$wrapOsx")
+def wrapWinPath = file("$buildDir/jwrappers/$wrapWin")
+
+task wrapInstaller(type: Copy) {
+    dependsOn tasks.installer, configurations.wrapTar, configurations.wrapZip
+
+    from zipTree(configurations.wrapZip.singleFile)
+    into "$buildDir/jwrappers"
```
Comment (mattmess1221) [inline build.gradle:231]: You should use the temporaryDir instead. That's what this type of thing is meant for.

**Review by mattmess1221 — COMMENTED** [review]
A couple things. Probably no need to keep the commented  launch4j stuff in there either.

**Comment by 0xilly** [comment]
@mezz Sure :D,  A while ago Lex asked for a java 8 wrapper that could be controlled by forge using Mojangs JRE. After talking more with Lex he requested a that the wrapper have the ability to run any jar that has been appended to the end of it, so we could use the wrapper with the forge installer and replace launch4j.

File: .travis.yml [hunk 9]
```diff
@@ -23,7 +23,7 @@ install:
   - ./gradlew --refresh-dependencies
   - ./gradlew ciWriteBuildNumber
 script:
-  - travis_wait ./gradlew build -x createExe -S
+  - travis_wait ./gradlew build -S
```
Comment (mezz) [inline .travis.yml:26]: @Illyohs The `-x` there means to skip the `createExe` step. 
It seems your commit message is the opposite of code change. Which is correct?

Comment (0xilly) [reply .travis.yml:26]: Correct, `createExe` no longer exists in the build process so in order to fix the ci build I removed it

Comment (mezz) [reply .travis.yml:26]: Ok thanks, that makes sense.

**Comment by 0xilly** [comment]
Sorry for the late push I forgot to push the mac packing

**Comment by mattmess1221** [comment]
Didn't I mention earlier about using groovy expanded properties in the app.xml instead of the `@IDENTIFIER@`?

**Comment by 0xilly** [comment]
damn it fixing the conflict now

**Comment by 0xilly** [comment]
Should be ready to merge into master

**Comment by mezz** [comment]
@Illyohs please remove the `jsons/1.11.2.json` file changes, it doesn't belong in this PR and causes a conflict.
You should discuss this with Lex to get it merged.

**Comment by LexManos** [comment]
This is also not currently in an acceptable state, The entire point of this PR is to make Macs install easier, which doesn't work as is. You need admin privileges apparently.
As for windows, yes grabbing the JVM from mojang's official spots would be nice, but not strictly necessary. We've only had a few issues reported where people didn't have a acceptable JRE installed for L4J.

**Comment by 0xilly** [comment]
On Sun, 2017-05-21 at 12:40 -0700, LexManos wrote:
> This is also not currently in an acceptable state, The entire point
> of this PR is to make Macs install easier, which doesn't work as is.
> You need admin privileges apparently.
Odd Ill test this once I get a mac again. The mac build shouldn't work 
 due to UPX breaking the OSX binary. I need to build UPX to see if it's
fixed(this was the original blocker)

> As for windows, yes grabbing the JVM from mojang's official spots
> would be nice, but not strictly necessary. We've only had a few 
Is forge planning to also host the JVM? I currently only grab it if 
it exists in {MC_DIR}/runtime/{JRE_VERSION}/

> issues reported where people didn't have a acceptable JRE installed
> for L4J.
This is a new one I'll look into it could do we have crash logs?

I get you a re-packaged UPX later this week(hopefully Tuesday) once I
get back in state.
--
Anderson

**Comment by 0xilly** [comment]
Closing will reopen targeting 1.12.x once Lex's issues are fully addressed.

**Pull request closed without merging**